File: vault-sdk-ios/Sources/Vault/Auth/AuthService.swift
import Foundation
import Combine

/// The authentication service for managing user sign-in and sign-up.
///
/// Use this class to authenticate users with various methods including
/// email/password, magic links, and biometric authentication.
public actor VaultAuth {
    
    // MARK: - Properties
    
    private let apiClient: APIClient
    private let tokenStore: TokenStore
    
    // MARK: - Initialization
    
    init(apiClient: APIClient, tokenStore: TokenStore) {
        self.apiClient = apiClient
        self.tokenStore = tokenStore
    }
    
    // MARK: - Email/Password Authentication
    
    /// Signs in a user with email and password.
    ///
    /// - Parameters:
    ///   - email: The user's email address
    ///   - password: The user's password
    /// - Returns: The session containing the user and tokens
    /// - Throws: `VaultError` if authentication fails
    public func signIn(email: String, password: String) async throws -> VaultSession {
        // Validate inputs
        guard email.isValidEmail else {
            throw ValidationError.invalidEmail
        }
        guard !password.isEmpty else {
            throw ValidationError.requiredField("Password")
        }
        
        let request = SignInRequest(email: email, password: password)
        let response: AuthResponse = try await apiClient.post(
            path: "/v1/auth/signin",
            body: request,
            requiresAuth: false
        )
        
        // Store tokens
        await tokenStore.setTokens(
            accessToken: response.accessToken,
            refreshToken: response.refreshToken
        )
        
        Vault.shared.logger.log("User signed in: \(response.user.id)")
        
        return Vault.shared.session
    }
    
    /// Signs up a new user with email and password.
    ///
    /// - Parameters:
    ///   - email: The user's email address
    ///   - password: The user's password
    ///   - firstName: Optional first name
    ///   - lastName: Optional last name
    /// - Returns: The session containing the user and tokens
    /// - Throws: `VaultError` if sign-up fails
    public func signUp(
        email: String,
        password: String,
        firstName: String? = nil,
        lastName: String? = nil
    ) async throws -> VaultSession {
        // Validate inputs
        guard email.isValidEmail else {
            throw ValidationError.invalidEmail
        }
        guard password.count >= 8 else {
            throw ValidationError.tooShort(field: "Password", minLength: 8)
        }
        
        let request = SignUpRequest(
            email: email,
            password: password,
            firstName: firstName,
            lastName: lastName
        )
        let response: AuthResponse = try await apiClient.post(
            path: "/v1/auth/signup",
            body: request,
            requiresAuth: false
        )
        
        // Store tokens
        await tokenStore.setTokens(
            accessToken: response.accessToken,
            refreshToken: response.refreshToken
        )
        
        Vault.shared.logger.log("User signed up: \(response.user.id)")
        
        return Vault.shared.session
    }
    
    /// Signs in with a biometric key.
    ///
    /// This requires that a biometric key has been previously generated and registered.
    ///
    /// - Returns: The session if authentication succeeds
    /// - Throws: `VaultError` if biometric authentication fails
    public func signInWithBiometric() async throws -> VaultSession {
        let biometric = Vault.shared.biometric
        
        guard biometric.isAvailable else {
            throw VaultError.biometricNotAvailable
        }
        
        // Get the stored private key
        guard let privateKey = try await biometric.getStoredKey() else {
            throw VaultError.keyNotFound
        }
        
        // Generate a challenge from the server
        let challenge: ChallengeResponse = try await apiClient.post(
            path: "/v1/auth/biometric/challenge",
            body: ["tenant_id": Vault.shared.configuration?.tenantId],
            requiresAuth: false
        )
        
        // Sign the challenge with the private key
        let signature = try await biometric.signChallenge(challenge.challenge, with: privateKey)
        
        // Verify the signature with the server
        let verifyRequest = BiometricVerifyRequest(
            challengeId: challenge.id,
            signature: signature,
            keyId: challenge.keyId
        )
        let response: AuthResponse = try await apiClient.post(
            path: "/v1/auth/biometric/verify",
            body: verifyRequest,
            requiresAuth: false
        )
        
        // Store tokens
        await tokenStore.setTokens(
            accessToken: response.accessToken,
            refreshToken: response.refreshToken
        )
        
        Vault.shared.logger.log("User signed in with biometric: \(response.user.id)")
        
        return Vault.shared.session
    }
    
    /// Registers biometric authentication for the current user.
    ///
    /// This generates a new key pair in the Secure Enclave and registers the public key
    /// with the Vault server.
    ///
    /// - Throws: `VaultError` if registration fails
    public func registerBiometric() async throws {
        let biometric = Vault.shared.biometric
        
        guard biometric.isAvailable else {
            throw VaultError.biometricNotAvailable
        }
        
        // Generate a new key pair in Secure Enclave
        let publicKey = try await biometric.generateAndStoreKey()
        
        // Get the public key data
        guard let publicKeyData = SecKeyCopyExternalRepresentation(publicKey, nil) else {
            throw VaultError.keyGenerationFailed("Failed to export public key")
        }
        
        // Register with server
        let request = RegisterBiometricRequest(
            publicKey: (publicKeyData as Data).base64EncodedString(),
            deviceName: await UIDevice.current.name
        )
        let _: EmptyResponse = try await apiClient.post(
            path: "/v1/auth/biometric/register",
            body: request
        )
        
        Vault.shared.logger.log("Biometric authentication registered")
    }
    
    /// Unregisters biometric authentication for the current user.
    ///
    /// - Throws: `VaultError` if unregistration fails
    public func unregisterBiometric() async throws {
        // Delete from server
        let _: EmptyResponse = try await apiClient.delete(path: "/v1/auth/biometric")
        
        // Delete local key
        try await Vault.shared.biometric.deleteStoredKey()
        
        Vault.shared.logger.log("Biometric authentication unregistered")
    }
    
    // MARK: - Password Management
    
    /// Requests a password reset email.
    ///
    /// - Parameter email: The email address to send the reset link to
    /// - Throws: `VaultError` if the request fails
    public func requestPasswordReset(email: String) async throws {
        guard email.isValidEmail else {
            throw ValidationError.invalidEmail
        }
        
        let request = PasswordResetRequest(email: email)
        let _: EmptyResponse = try await apiClient.post(
            path: "/v1/auth/password/reset",
            body: request,
            requiresAuth: false
        )
        
        Vault.shared.logger.log("Password reset requested for: \(email)")
    }
    
    /// Updates the current user's password.
    ///
    /// - Parameters:
    ///   - currentPassword: The current password
    ///   - newPassword: The new password
    /// - Throws: `VaultError` if the update fails
    public func updatePassword(currentPassword: String, newPassword: String) async throws {
        guard newPassword.count >= 8 else {
            throw ValidationError.tooShort(field: "New password", minLength: 8)
        }
        
        let request = UpdatePasswordRequest(
            currentPassword: currentPassword,
            newPassword: newPassword
        )
        let _: EmptyResponse = try await apiClient.post(
            path: "/v1/auth/password",
            body: request
        )
        
        Vault.shared.logger.log("Password updated")
    }
    
    // MARK: - Email Verification
    
    /// Verifies the user's email address.
    ///
    /// - Parameter token: The verification token from the email
    /// - Throws: `VaultError` if verification fails
    public func verifyEmail(token: String) async throws {
        let request = VerifyEmailRequest(token: token)
        let _: EmptyResponse = try await apiClient.post(
            path: "/v1/auth/verify-email",
            body: request,
            requiresAuth: false
        )
        
        Vault.shared.logger.log("Email verified")
    }
    
    /// Resends the verification email.
    ///
    /// - Throws: `VaultError` if the request fails
    public func resendVerificationEmail() async throws {
        let _: EmptyResponse = try await apiClient.post(
            path: "/v1/auth/resend-verification",
            body: EmptyRequest()
        )
        
        Vault.shared.logger.log("Verification email resent")
    }
    
    // MARK: - Session Management
    
    /// Refreshes the current session.
    ///
    /// - Returns: `true` if the session was refreshed successfully
    public func refreshSession() async -> Bool {
        guard let refreshToken = await tokenStore.getRefreshToken() else {
            return false
        }
        
        do {
            let request = RefreshTokenRequest(refreshToken: refreshToken)
            let response: TokenResponse = try await apiClient.post(
                path: "/v1/auth/refresh",
                body: request,
                requiresAuth: false
            )
            
            await tokenStore.setTokens(
                accessToken: response.accessToken,
                refreshToken: response.refreshToken
            )
            
            Vault.shared.logger.log("Session refreshed")
            return true
        } catch {
            Vault.shared.logger.logError("Failed to refresh session: \(error)")
            return false
        }
    }
}

// MARK: - Request/Response Types

internal struct AuthResponse: Decodable {
    let user: UserResponse
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int
}

internal struct UserResponse: Decodable {
    let id: String
    let email: String
    let firstName: String?
    let lastName: String?
    let emailVerified: Bool
    let createdAt: String
    let updatedAt: String
    let organizations: [OrganizationResponse]?
}

internal struct OrganizationResponse: Decodable {
    let id: String
    let name: String
    let slug: String
    let role: String
}

internal struct TokenResponse: Decodable {
    let accessToken: String
    let refreshToken: String
    let expiresIn: Int
}

internal struct ChallengeResponse: Decodable {
    let id: String
    let challenge: String
    let keyId: String
}

internal struct BiometricVerifyRequest: Encodable {
    let challengeId: String
    let signature: String
    let keyId: String
}

internal struct RegisterBiometricRequest: Encodable {
    let publicKey: String
    let deviceName: String
}

internal struct EmptyRequest: Encodable {}

// MARK: - String Extensions

extension String {
    var isValidEmail: Bool {
        let regex = "^[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$"
        return NSPredicate(format: "SELF MATCHES %@", regex).evaluate(with: self)
    }
}

File: fantasticauth-sdk-android/src/main/java/dev.fantasticauth/sdk/auth/AuthService.kt
package dev.vault.sdk.auth

import dev.vault.sdk.Vault
import dev.vault.sdk.network.APIClient
import dev.vault.sdk.network.VaultException
import dev.vault.sdk.session.TokenStore
import dev.vault.sdk.session.VaultSession
import dev.vault.sdk.user.User
import dev.vault.sdk.utils.VaultLogger
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Authentication service for Vault.
 * Handles email/password, magic link, and token-based authentication.
 */
class VaultAuth {
    
    private val apiClient = APIClient.instance
    private val tokenStore = TokenStore(Vault.context)
    
    /**
     * Sign in with email and password
     * 
     * @param email User's email address
     * @param password User's password
     * @param enableBiometric Whether to enable biometric for future sign-ins
     * @return Session with user and tokens
     * @throws VaultException if authentication fails
     */
    suspend fun signIn(
        email: String,
        password: String,
        enableBiometric: Boolean = false
    ): VaultSession.SessionData = withContext(Dispatchers.IO) {
        VaultLogger.d("Signing in user: $email")
        
        val request = SignInRequest(
            email = email,
            password = password,
            deviceInfo = DeviceInfo.collect()
        )
        
        val response = apiClient.post<SignInRequest, AuthResponse>(
            endpoint = "/auth/signin",
            body = request
        )
        
        response.toSessionData().also { session ->
            tokenStore.saveTokens(
                accessToken = session.accessToken,
                refreshToken = session.refreshToken
            )
            
            if (enableBiometric) {
                tokenStore.enableBiometricKey()
            }
            
            VaultLogger.i("User signed in: ${session.user.id}")
        }
    }
    
    /**
     * Sign up with email and password
     * 
     * @param email User's email address
     * @param password User's password
     * @param name User's full name
     * @return Session with user and tokens
     * @throws VaultException if signup fails
     */
    suspend fun signUp(
        email: String,
        password: String,
        name: String? = null
    ): VaultSession.SessionData = withContext(Dispatchers.IO) {
        VaultLogger.d("Signing up user: $email")
        
        val request = SignUpRequest(
            email = email,
            password = password,
            name = name,
            deviceInfo = DeviceInfo.collect()
        )
        
        val response = apiClient.post<SignUpRequest, AuthResponse>(
            endpoint = "/auth/signup",
            body = request
        )
        
        response.toSessionData().also { session ->
            tokenStore.saveTokens(
                accessToken = session.accessToken,
                refreshToken = session.refreshToken
            )
            VaultLogger.i("User signed up: ${session.user.id}")
        }
    }
    
    /**
     * Sign in with biometric (previously enrolled)
     * 
     * @return Session with user and tokens
     * @throws VaultException if biometric auth fails or not enrolled
     */
    suspend fun signInWithBiometric(): VaultSession.SessionData = withContext(Dispatchers.IO) {
        if (!tokenStore.isBiometricEnabled()) {
            throw VaultException(
                code = "biometric_not_enrolled",
                message = "Biometric authentication not enabled for this account"
            )
        }
        
        val biometricKey = tokenStore.getBiometricKey()
            ?: throw VaultException(
                code = "biometric_key_missing",
                message = "Biometric key not found"
            )
        
        VaultLogger.d("Signing in with biometric")
        
        val request = BiometricAuthRequest(
            keyId = biometricKey.keyId,
            signature = biometricKey.createSignature(),
            deviceInfo = DeviceInfo.collect()
        )
        
        val response = apiClient.post<BiometricAuthRequest, AuthResponse>(
            endpoint = "/auth/biometric",
            body = request
        )
        
        response.toSessionData().also { session ->
            tokenStore.saveTokens(
                accessToken = session.accessToken,
                refreshToken = session.refreshToken
            )
            VaultLogger.i("User signed in with biometric: ${session.user.id}")
        }
    }
    
    /**
     * Request magic link for passwordless sign-in
     * 
     * @param email User's email address
     * @param redirectUrl Optional redirect URL after magic link confirmation
     * @throws VaultException if request fails
     */
    suspend fun requestMagicLink(
        email: String,
        redirectUrl: String? = null
    ) = withContext(Dispatchers.IO) {
        VaultLogger.d("Requesting magic link for: $email")
        
        val request = MagicLinkRequest(
            email = email,
            redirectUrl = redirectUrl
        )
        
        apiClient.post<MagicLinkRequest, Unit>(
            endpoint = "/auth/magic-link",
            body = request
        )
        
        VaultLogger.i("Magic link requested for: $email")
    }
    
    /**
     * Sign in with magic link token
     * 
     * @param token Magic link token from email
     * @return Session with user and tokens
     * @throws VaultException if token is invalid
     */
    suspend fun signInWithMagicLink(token: String): VaultSession.SessionData = 
        withContext(Dispatchers.IO) {
            VaultLogger.d("Signing in with magic link")
            
            val request = MagicLinkSignInRequest(
                token = token,
                deviceInfo = DeviceInfo.collect()
            )
            
            val response = apiClient.post<MagicLinkSignInRequest, AuthResponse>(
                endpoint = "/auth/magic-link/verify",
                body = request
            )
            
            response.toSessionData().also { session ->
                tokenStore.saveTokens(
                    accessToken = session.accessToken,
                    refreshToken = session.refreshToken
                )
                VaultLogger.i("User signed in with magic link: ${session.user.id}")
            }
        }
    
    /**
     * Send password reset email
     * 
     * @param email User's email address
     * @throws VaultException if request fails
     */
    suspend fun forgotPassword(email: String) = withContext(Dispatchers.IO) {
        VaultLogger.d("Requesting password reset for: $email")
        
        val request = ForgotPasswordRequest(email = email)
        
        apiClient.post<ForgotPasswordRequest, Unit>(
            endpoint = "/auth/forgot-password",
            body = request
        )
        
        VaultLogger.i("Password reset requested for: $email")
    }
    
    /**
     * Reset password with token
     * 
     * @param token Reset token from email
     * @param newPassword New password
     * @throws VaultException if token is invalid
     */
    suspend fun resetPassword(
        token: String,
        newPassword: String
    ) = withContext(Dispatchers.IO) {
        VaultLogger.d("Resetting password")
        
        val request = ResetPasswordRequest(
            token = token,
            newPassword = newPassword
        )
        
        apiClient.post<ResetPasswordRequest, Unit>(
            endpoint = "/auth/reset-password",
            body = request
        )
        
        VaultLogger.i("Password reset successful")
    }
    
    /**
     * Verify email address
     * 
     * @param token Verification token from email
     * @throws VaultException if token is invalid
     */
    suspend fun verifyEmail(token: String) = withContext(Dispatchers.IO) {
        VaultLogger.d("Verifying email")
        
        val request = VerifyEmailRequest(token = token)
        
        apiClient.post<VerifyEmailRequest, Unit>(
            endpoint = "/auth/verify-email",
            body = request
        )
        
        VaultLogger.i("Email verified")
    }
    
    /**
     * Resend verification email
     * 
     * @param email User's email address
     * @throws VaultException if request fails
     */
    suspend fun resendVerification(email: String) = withContext(Dispatchers.IO) {
        VaultLogger.d("Resending verification email for: $email")
        
        val request = ResendVerificationRequest(email = email)
        
        apiClient.post<ResendVerificationRequest, Unit>(
            endpoint = "/auth/resend-verification",
            body = request
        )
        
        VaultLogger.i("Verification email resent for: $email")
    }
    
    /**
     * Sign out the current user
     * Clears local tokens and optionally notifies the server
     * 
     * @param revokeAll Whether to revoke all sessions (default: false)
     */
    suspend fun signOut(revokeAll: Boolean = false) = withContext(Dispatchers.IO) {
        VaultLogger.d("Signing out user")
        
        try {
            if (revokeAll) {
                apiClient.post<Unit, Unit>(endpoint = "/auth/signout/all", body = Unit)
            } else {
                apiClient.post<Unit, Unit>(endpoint = "/auth/signout", body = Unit)
            }
        } catch (e: Exception) {
            VaultLogger.w("Failed to notify server of signout: ${e.message}")
        } finally {
            tokenStore.clearTokens()
            VaultLogger.i("User signed out")
        }
    }
    
    // Data classes for requests and responses
    
    private data class SignInRequest(
        val email: String,
        val password: String,
        val deviceInfo: DeviceInfo
    )
    
    private data class SignUpRequest(
        val email: String,
        val password: String,
        val name: String?,
        val deviceInfo: DeviceInfo
    )
    
    private data class BiometricAuthRequest(
        val keyId: String,
        val signature: String,
        val deviceInfo: DeviceInfo
    )
    
    private data class MagicLinkRequest(
        val email: String,
        val redirectUrl: String?
    )
    
    private data class MagicLinkSignInRequest(
        val token: String,
        val deviceInfo: DeviceInfo
    )
    
    private data class ForgotPasswordRequest(val email: String)
    
    private data class ResetPasswordRequest(
        val token: String,
        val newPassword: String
    )
    
    private data class VerifyEmailRequest(val token: String)
    
    private data class ResendVerificationRequest(val email: String)
    
    private data class AuthResponse(
        val user: User,
        val accessToken: String,
        val refreshToken: String,
        val expiresIn: Long,
        val tokenType: String = "Bearer"
    ) {
        fun toSessionData() = VaultSession.SessionData(
            user = user,
            accessToken = accessToken,
            refreshToken = refreshToken,
            expiresIn = expiresIn,
            tokenType = tokenType
        )
    }
}
